rust/src/shards/gui/layouts/scroll_area.rs

```diff
@@ -0,0 +1,179 @@
+    Self {
+      parents,
+      requiring: Vec::new(),
+      contents: ShardsVar::default(),
+      horizontal: ParamVar::new(false.into()),
+      vertical: ParamVar::new(true.into()),
+    }
+  }
+}
+
+impl Shard for ScrollArea {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.ScrollArea")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.ScrollArea-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.ScrollArea"
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&AREA_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => self.contents.set_param(value),
+      1 => Ok(self.horizontal.set_param(value)),
+      2 => Ok(self.vertical.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.contents.get_param(),
+      1 => self.horizontal.get_param(),
+      2 => self.vertical.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    let exp_info = ExposedInfo {
+      exposedType: EGUI_UI_SEQ_TYPE,
+      name: self.parents.get_name(),
+      help: cstr!("The parent UI objects.").into(),
+      ..ExposedInfo::default()
+    };
+    self.requiring.push(exp_info);
+
+    Some(&self.requiring)
+  }
+
+  fn hasCompose() -> bool {
+    true
+  }
+
+  fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
+    if !self.contents.is_empty() {
+      let outputType = self.contents.compose(&data)?;
+      return Ok(outputType);
+    }
+
+    Ok(data.inputType)
+  }
+
+  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+    self.parents.warmup(ctx);
+    if !self.contents.is_empty() {
+      self.contents.warmup(ctx)?;
+    }
+    self.horizontal.warmup(ctx);
+    self.vertical.warmup(ctx);
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.vertical.cleanup();
+    self.horizontal.cleanup();
+    if !self.contents.is_empty() {
+      self.contents.cleanup();
+    }
+    self.parents.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    if self.contents.is_empty() {
+      return Ok(*input);
+    }
+
+    if let Some(ui) = util::get_current_parent(*self.parents.get())? {
+      let output = egui::ScrollArea::new([
+        self.horizontal.get().try_into()?,
+        self.vertical.get().try_into()?,
+      ])
+      .show(ui, |ui| {
+        util::activate_ui_contents(context, input, ui, &mut self.parents, &mut self.contents)
+      })
+      .inner?;
+
+      Ok(output)
```

Comment:
should be `*input`

Reply:
Propagation of the output of the inner shards (see comment on collapsing header and README).